File: generator/src/main/groovy/ch/bailu/gtk/model/CType.java
package ch.bailu.gtk.model;

import java.util.regex.Pattern;

public class CType {

    private final static Pattern P_CONST = Pattern.compile("^const [A-Za-z]*");
    private final static Pattern P_POINTER = Pattern.compile(".*[A-Za-z]\\*$");

    final boolean isPointer;
    final boolean isConst;

    private final String type;

    public CType(String type) {
        if (type == null) {
            type = "void*";
        }


        this.type = type;

        isPointer = P_POINTER.matcher(this.type).find();
        isConst = P_CONST.matcher(this.type).find();
    }

    public boolean contains(String type) {
        return this.type.contains(type);
    }

    public boolean isSinglePointer() {
        return isPointer;
    }
    public boolean isConst() {
        return isConst;
    }
    public String getName() {
        return type;
    }
}

File: generator/src/main/kotlin/ch/bailu/gtk/model/type/CType.kt
package ch.bailu.gtk.model.type

import ch.bailu.gtk.log.DebugPrint
import java.util.regex.Pattern

class CType(type: String) : Type() {
    companion object {
        private val P_CONST = Pattern.compile("^const [A-Za-z]+")
        private val P_POINTER = Pattern.compile(".*[A-Za-z]+\\*$")
        private val P_NOPOINTER = Pattern.compile(".*[A-Za-z]+$")
    }

    val type = if ("" == type) "void*" else type

    val isSinglePointer = P_POINTER.matcher(this.type).find()
    val isConst         = P_CONST.matcher(this.type).find()
    val isDirectType    = P_NOPOINTER.matcher(this.type).find()

    operator fun contains(type: String): Boolean {
        return this.type.contains(type)
    }

    override fun getDebugIdentifier(): String {
        return "c"
    }

    override fun toString(): String {
        return DebugPrint.colon(this, type)
    }
}
